### PR TITLE
Refine the size() calculation of accumulator

### LINKED ISSUE #5903: Refine the size() calculation of accumulator
### Is your feature request related to a problem or challenge?

_No response_

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #5903.

# Rationale for this change

From the flame graph generated by 
`
CARGO_PROFILE_RELEASE_DEBUG=true cargo flamegraph --no-inline --freq 500 --bin tpch -- benchmark datafusion --path ./data-parquet/ --format parquet --partitions 1 -q 17 --iterations 10
`
flamegraph-main
There are around 4% spent on calling the size() of accumulator, which can be improved.

# What changes are included in this PR?

- Avoid unnecessary calling of `ScalarValue`'s size().
- Avoid duplicated calculation of the initial size of `accumulator_set`.
- Pull out the mode check to the top level to avoid if-else judgement for every group.

After applying this PR, from the flame graph, we can see there's almost no cost on the size() of accumulator. The benchmark result is improved from 11s to 10.5s.
flamegraph-5903

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/aggregate/sum.rs [hunk 1]
```diff
@@ -266,7 +266,7 @@ impl Accumulator for SumAccumulator {
     }
 
     fn size(&self) -> usize {
-        std::mem::size_of_val(self) - std::mem::size_of_val(&self.sum) + self.sum.size()
+        std::mem::size_of_val(self)
     }
```
Comment (mingmwang) [inline datafusion/physical-expr/src/aggregate/sum.rs:270]: Will this size change after the AvgAccumulator /SumAccumulator struct is initialized?

Comment (mingmwang) [reply datafusion/physical-expr/src/aggregate/sum.rs:270]: I would suggest rename the `size` method to `size_in_bytes`

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 2]
```diff
@@ -514,6 +580,19 @@ impl GroupedHashAggregateStream {
...
+    groups_with_rows.iter().fold(0usize, |acc, group_idx| {
+        let group_state = &row_group_states[*group_idx];
+        group_state
+            .accumulator_set
+            .iter()
+            .fold(acc, |acc, accumulator| acc + accumulator.size())
+    })
+}
+
 /// The state that is built for each output group.
```
Comment (mingmwang) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:596]: I think the logic is quite complex to collect the memory size of the accumulators, maybe the computation is more than the real useful aggregations

File: datafusion/physical-expr/src/aggregate/sum.rs (hunk 1 above)
Comment (yahoNanJing) [reply datafusion/physical-expr/src/aggregate/sum.rs:270]: It may be changed. For example, to deal with null.

Comment (mingmwang) [reply datafusion/physical-expr/src/aggregate/sum.rs:270]: I think not. Option is of Enum type.

Comment (yahoNanJing) [reply datafusion/physical-expr/src/aggregate/sum.rs:270]: From the flamegraph, it seems it's not a bottleneck now.

**Comment by yahoNanJing** [comment]
Hi @Dandandan, @alamb, could you help review this PR?

**Comment by alamb** [comment]
I will try and review this carefully later today

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 3]
```diff
@@ -419,92 +423,154 @@ impl GroupedHashAggregateStream {
...
-                                | AggregateMode::Final => {
+                                })
+                                // 2.5
+                                .and({
+                                    group_state.indices.clear();
+                                    Ok(())
+                                })
+                        })?;
+                }
+                AggregateMode::FinalPartitioned | AggregateMode::Final => {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:501]: Maybe github is rendering the diff confusingly, but this seems like a significant amount of new code

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/aggregates/row_hash.rs:596]: Yes, I agree the calculation for the memory size is complicated. 

I think @tustvold  has been thinking about how to improve performance in this area, but I am not sure how far he has gotten

In general, managing individual allocations (and then accounting for their sizes) for each group is a significant additional overhead for grouping.

**Review by alamb — COMMENTED** [review]
Thanks @yahoNanJing 

The idea of caching the accumulator sizes seems like a good one.

I am not sure about the changes to `Sum` and `Avg` accumulators

Maybe we need to take a step back and figure out how to improve grouping performance more holistically rather than trying to do a special optimization on allocation accounting 🤔 

For example https://github.com/apache/arrow-datafusion/issues/4973

**Comment by alamb** [comment]
I am trying to clean up outstanding PRs and I came across this one. What shall we do with it @yahoNanJing  -- should we pursue getting it merged?

**Comment by alamb** [comment]
Since this has been open for more than a year, closing it down. Feel free to reopen if/when you keep working on it.

**Pull request closed without merging**